### PR TITLE
Using the new options for image-scanning options

### PR DESCRIPTION
**Pull request by enoodle**
Based on https://github.com/ManageIQ/manageiq/pull/15398

This will use the options field for per-provider configuration in image scanning.
Also, This will try to use the http_proxy set in the options to connect to the provider instead of the one defined in the settings file. Originally it was planned to have `https_proxy` and `no_proxy` options as well for the connection to the provider but as they are not used currently I left them out.

### CODE REVIEW

**Comment by simon3z** [comment]
@enoodle nice! I think this would do it.

File: app/models/manageiq/providers/kubernetes/container_manager/options.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            :global_default => Settings.ems.ems_kubernetes.image_inspector_registry,
+          },
+          :image_tag   => {
+            :label          => N_('Image-Inspector Tag'),
+            :help_text      => N_('Image-Inspector image tag'),
+            :global_default => ManageIQ::Providers::Kubernetes::ContainerManager::Scanning::Job::INSPECTOR_IMAGE_TAG,
+          },
+          :cve_url     => {
+            :label     => N_('CVE location'),
+            :help_text => N_('Enables defining a URL for XCCDF file instead of accessing the Internet'),
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager/options.rb]: I remember something about this being a prefix  and not url? 

would it be possible to add an example to each value, even only a comment?

I'm thinking this would great if I need to use one of these I could look here and see what I need.

**Comment by enoodle** [comment]
@moolitayer I can also have the example in the `help_text` that I plan to present in the UI (when you hover over the field)

**Comment by moolitayer** [comment]
> @moolitayer I can also have the example in the help_text that I plan to present in the UI (when you hover over the field)

:+1: That would be much better :hot_pepper:

**Comment by enoodle** [comment]
@moolitayer done.

File: app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb [hunk 2]
```diff
@@ -348,6 +347,7 @@ def inspector_admin_secret
   end
 
   def pod_definition(inspector_admin_secret_name)
+    @ems_options = ext_management_system.options[:image_inspector_options]
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb]: Can we please have ems_options as a method and not an instance variable?
I might be reading this wrong but the usage of `@ems_options[:registry]` in line 429 looks like a side effect of calling this method first. (you can memoize it if you want to save lookups)

Comment (enoodle) [inline app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb]: @moolitayer I wanted to prevent accessing the DB multiple times to fetch this value. I think maybe to move it to the `initializing` function. what do you think?

File: app/models/manageiq/providers/kubernetes/container_manager/options.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            :global_default => Settings.ems.ems_kubernetes.image_inspector_registry,
+          },
+          :image_tag   => {
+            :label          => N_('Image-Inspector Tag'),
+            :help_text      => N_('Image-Inspector image tag. example: 2.1'),
+            :global_default => ManageIQ::Providers::Kubernetes::ContainerManager::Scanning::Job::INSPECTOR_IMAGE_TAG,
+          },
+          :cve_url     => {
+            :label     => N_('CVE location'),
+            :help_text => N_('Enables defining a URL path prefix for XCCDF file instead of accessing the default location.
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager/options.rb]: :+1:

File: app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb [hunk 4]
```diff
@@ -347,6 +346,10 @@ def inspector_admin_secret
     return nil
   end
 
+  def ems_options
+    @provider_options ||= ext_management_system.options[:image_inspector_options]
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb]: :+1:

File: app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb [hunk 5]
```diff
@@ -425,17 +429,25 @@ def add_secret_to_pod_def(pod_def, inspector_admin_secret_name)
...
   end
 
   def inspector_proxy_env_variables
-    settings = ext_management_system.custom_attributes
-    settings.where(:section => ATTRIBUTE_SECTION,
-                   :name    => PROXY_ENV_VARIABLES).each_with_object([]) do |att, env|
-      env << {:name  => att.name.upcase,
-              :value => att.value} unless att.value.blank?
+    PROXY_ENV_VARIABLES.each_with_object([]) do |var_name, env|
+      if ems_options.keys.include?(var_name.to_sym)
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb]: `ems_options.key?(var_name.to_sym)`

File: app/models/manageiq/providers/kubernetes/container_manager_mixin.rb [hunk 6]
```diff
@@ -108,11 +108,12 @@ def ssl_cert_store(endpoint = default_endpoint)
...
-      :port     => options[:port] || port,
-      :user     => options[:user] || authentication_userid(options[:auth_type]),
-      :pass     => options[:pass] || authentication_password(options[:auth_type]),
-      :bearer   => options[:bearer] || authentication_token(options[:auth_type] || 'bearer'),
+      :hostname    => options[:hostname] || address,
+      :port        => options[:port] || port,
+      :user        => options[:user] || authentication_userid(options[:auth_type]),
+      :pass        => options[:pass] || authentication_password(options[:auth_type]),
+      :bearer      => options[:bearer] || authentication_token(options[:auth_type] || 'bearer'),
+      :http_proxy  => self.options.fetch_path(:proxy_settings, :http_proxy),
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager_mixin.rb]: ems_options.fetch_path maybe? `self.options` is nil according to Travis

**Comment by moolitayer** [comment]
cc @cben @simon3z

**Comment by simon3z** [comment]
@enoodle can you fix travis?

**Comment by enoodle** [comment]
@moolitayer I added tests and now travis is green. PTAL

File: app/models/manageiq/providers/kubernetes/container_manager/options.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
+module ManageIQ::Providers::Kubernetes::ContainerManager::Options
+  extend ActiveSupport::Concern
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/options.rb]: do you need ActiveSupport::Concern?  what does it do here?  (`extend` into an `extend`ed module is too much for my brain ;-)

I think you don't need it, this simply has methods that become class methods due to `extend ...ContainerManager::Options`.
or if you want to use ActiveSupport::Concern idiomatically, switch to `include`ing this module and put the methods under `ClassMethods`.

File: app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb [hunk 8]
```diff
@@ -347,6 +346,10 @@ def inspector_admin_secret
     return nil
   end
 
+  def ems_options
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb]: consider renaming to `ems_image_inspector_options` (or something intermediate).
because `ems_options[:registry]` sounds like these are top-level options on the ems.

File: app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb [hunk 9]
```diff
@@ -425,17 +429,24 @@ def add_secret_to_pod_def(pod_def, inspector_admin_secret_name)
...
   end
 
   def inspector_proxy_env_variables
-    settings = ext_management_system.custom_attributes
-    settings.where(:section => ATTRIBUTE_SECTION,
-                   :name    => PROXY_ENV_VARIABLES).each_with_object([]) do |att, env|
-      env << {:name  => att.name.upcase,
-              :value => att.value} unless att.value.blank?
+    PROXY_ENV_VARIABLES.each_with_object([]) do |var_name, env|
+      next unless ems_options.key?(var_name.to_sym)
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb]: not sure it matters, but previous code checking `.blank?` allowed "unsetting" by setting to empty string, this requires actually unsetting the option — is that even possible with the new per-EMS options?

File: app/models/manageiq/providers/kubernetes/container_manager_mixin.rb [hunk 10]
```diff
@@ -108,11 +108,12 @@ def ssl_cert_store(endpoint = default_endpoint)
...
-      :port     => options[:port] || port,
-      :user     => options[:user] || authentication_userid(options[:auth_type]),
-      :pass     => options[:pass] || authentication_password(options[:auth_type]),
-      :bearer   => options[:bearer] || authentication_token(options[:auth_type] || 'bearer'),
+      :hostname    => options[:hostname] || address,
+      :port        => options[:port] || port,
+      :user        => options[:user] || authentication_userid(options[:auth_type]),
+      :pass        => options[:pass] || authentication_password(options[:auth_type]),
+      :bearer      => options[:bearer] || authentication_token(options[:auth_type] || 'bearer'),
+      :http_proxy  => self.options ? self.options.fetch_path(:proxy_settings, :http_proxy) : nil,
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager_mixin.rb]: why `:proxy_settings, :http_proxy` here and just `:http_proxy` in `kubernetes_connect` ?

Comment (enoodle) [inline app/models/manageiq/providers/kubernetes/container_manager_mixin.rb]: `kubernetes_connect` gets its options from here, calling `raw_connect` will eventually `kubernetes_connect`. I put them in `http_proxy` so no need for the `proxy_settings`. Openshift calls `kubernetes_connect` as well after adding its own options.

File: app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb (hunk 9 above)
Comment (enoodle) [inline app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb]: It is possible to totally unset an option if you set the options without the ones you don't want:
`ems.update(:options => {})` or `ems.update(:options => {:image_inspector_options => {:http_proxy => "my_proxy"}})`. The options are only set from the UI [1] so it will be up to that piece of code to make sure that empty values are not set.
I wonder if it is smart to add tests for the empty option case making sure that they are treated as unset. (then making sure it is so here) - This will prevent us from assigning them the empty value. WDYT?

[1] https://github.com/ManageIQ/manageiq-ui-classic/pull/1652/files#diff-f402e411cd6dc158c724b14d72ead612R445

File: app/models/manageiq/providers/kubernetes/container_manager.rb [hunk 11]
```diff
@@ -14,6 +14,7 @@ class ManageIQ::Providers::Kubernetes::ContainerManager < ManageIQ::Providers::C
   require_nested :Scanning
 
   include ManageIQ::Providers::Kubernetes::ContainerManagerMixin
+  include ManageIQ::Providers::Kubernetes::ContainerManager::Options
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager.rb]: :+1:

Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager.rb]: @enoodle @cben didn't we miss a require nested of the new file here?
